### ISSUE TITLE
Astropy 1.3 fails to open FITS files from MOSFIRE

### ISSUE THREAD

**Issue by joshwalawender** [astropy/astropy#5807]
Version 1.3 of astropy is failing to open FITS files from the MOSFIRE instrument at Keck.  The files pass a `fitsverify` check and open fine in astropy v1.2.1 (checked on both a mac and linux machine), but they open as a `_CorruptedHDU` in astropy 1.3.

This may have something to do with the format of the MOSFIRE files (the first HDU is image data, but there are 4 subsequent TableHDUs.

Here's the result when comparing astropy v1.3 and v1.2.1 on the same machine:

```
In [1]: import sys

In [2]: sys.version
Out[2]: '3.5.2 |Anaconda custom (x86_64)| (default, Jul  2 2016, 17:52:12) \n[GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)]'

In [3]: import astropy

In [4]: astropy.__version__
Out[4]: '1.3'

In [5]: from astropy.io import fits

In [6]: hdul = fits.open('m120910_0163.fits', 'readonly')
WARNING: An exception occurred matching an HDU header to the appropriate HDU type:  [astropy.io.fits.hdu.base]
WARNING: The HDU will be treated as corrupted. [astropy.io.fits.hdu.base]

In [7]: hdul
Out[7]: [<astropy.io.fits.hdu.base._CorruptedHDU object at 0x10a1e9080>]
```

and again with v1.2.1

```
In [1]: import sys

In [2]: sys.version
Out[2]: '3.5.2 |Anaconda custom (x86_64)| (default, Jul  2 2016, 17:52:12) \n[GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)]'

In [3]: import astropy

In [4]: astropy.__version__
Out[4]: '1.2.1'

In [5]: from astropy.io import fits

In [6]: hdul = fits.open('m120910_0163.fits', 'readonly')

In [7]: hdul
Out[7]:
[<astropy.io.fits.hdu.image.PrimaryHDU at 0x10a20f668>,
 <astropy.io.fits.hdu.table.TableHDU at 0x10a2256d8>,
 <astropy.io.fits.hdu.table.TableHDU at 0x10a229c18>,
 <astropy.io.fits.hdu.table.TableHDU at 0x10a22de48>,
 <astropy.io.fits.hdu.table.TableHDU at 0x10a234f98>]
```

A sample file can be downloaded here:
https://dl.dropboxusercontent.com/u/17112464/m120910_0163.fits
but the problem has been confirmed to exist with other MOSFIRE files as well.

-Josh

P.S.  Ran `fitsverify` here: https://fits.gsfc.nasa.gov/fits_verify.html

The tail of the result is:
```
++++++++++++++++++++++ Error Summary  ++++++++++++++++++++++
 
 HDU#  Name (version)       Type             Warnings  Errors
 1                          Primary Array    0         0     
 2     Target_List          ASCII Table      0         0     
 3     Science_Slit_List    ASCII Table      0         0     
 4     Mechanical_Slit_List ASCII Table      0         0     
 5     Alignment_Slit_List  ASCII Table      0         0     
 
**** Verification found 0 warning(s) and 0 error(s). ****
```

**Comment by astrofrog** [comment]
It looks like the culprit is https://github.com/astropy/astropy/pull/4581 (based on bisection)

**Comment by bsipocz** [comment]
I think it's probably been already fixed, on master I got the same as @joshwalawender with 1.2.1:

```
In [3]: In [1]: import sys
   ...: 
   ...: In [2]: sys.version
   ...: 
   ...: 
   ...: In [3]: import astropy
   ...: 
   ...: In [4]: astropy.__version__
   ...: 
Out[3]: '2.0.dev12241'

In [4]: In [5]: from astropy.io import fits
   ...: 
   ...: In [6]: hdul = fits.open('m120910_0163.fits', 'readonly')
   ...: 

In [5]: In [7]: hdul
Out[5]: 
[<astropy.io.fits.hdu.image.PrimaryHDU at 0x1121de898>,
 <astropy.io.fits.hdu.table.TableHDU at 0x1121f9908>,
 <astropy.io.fits.hdu.table.TableHDU at 0x1121fce48>,
 <astropy.io.fits.hdu.table.TableHDU at 0x1122017b8>,
 <astropy.io.fits.hdu.table.TableHDU at 0x1122065c0>]
```

**Comment by bsipocz** [comment]
(I need to find which PR did the fix though, to make sure it's backported).

**Comment by bsipocz** [comment]
Ignore my previous comment, I was on the wrong branch.

**Comment by bsipocz** [comment]
Fix is on its way. However it seems to me that the issue is originating from the fact that the header keyword ``GROUPS`` doesn't follow the standard, as it's value is 2 (``GROUPS  =                    2 / Number of sampling groups``).

The standard (from here: http://heasarc.gsfc.nasa.gov/docs/fcg/standard_dict.html) says this:

```
KEYWORD:   GROUPS
REFERENCE: FITS Standard 
STATUS:    mandatory
HDU:       groups
VALUE:     logical
COMMENT:   indicates random groups structure
DEFINITION: The value field shall contain the logical constant T.  The
value T associated with this keyword implies that random groups records
are present.
```

**Comment by mhvk** [comment]
@joshwalawender - would you be able to notify the people with MOSFIRE that their fits files do not confirm to the standard? While we were interpreting it too strictly, it would still be better if the files are written correctly to start with!

**Comment by joshwalawender** [comment]
I will see if the header can be changed for subsequent files, however, I'd like to find a solution for existing files if possible.

I tried a simple test of reading in the HDUList, editing the GROUPS header value to be compliant, but I can't write out a valid file using astropy.  I also tried `fits.setval`, but the problem is preventing me from fixing the header in python.

Is there a workaround that I'm missing?

**Comment by pllim** [comment]
@joshwalawender , for now, try using the dev version of Astropy and see if it works again. The fix should be in 1.3.1.

**Comment by bsipocz** [comment]
@joshwalawender - For the long term I would suggest to try and change your ``GROUPS`` keyword to something else, at least based on the comment in your header, it seems to mean something different than the one in the description of the standard. 
As @pllim said the fix will be in 1.3.1, which is to be released soon(ish).

**Comment by joshwalawender** [comment]
Ok, thanks everyone!  I'm warning our users away from v1.3 until v1.3.1 is out, so far it looks as if only one user noticed the issue.  I'll start the process of changing the acquisition software to get the FITS headers written properly.
